From 0c8707f42ec095d7337da6f07c6c9f70535d33c0 Mon Sep 17 00:00:00 2001
From: Ben Wolsieffer <benwolsieffer@gmail.com>
Date: Sat, 8 Dec 2018 20:35:56 -0500
Subject: [PATCH] ASoC: sgtl5000: add I2S output source selection

Add mux for choosing the I2S output source: ADC, I2S input or DAP.
Previously, the I2S output source was always the ADC, making it
impossible to record using I2S while using DAP functions (automatic
volume control, bass enhance).

Signed-off-by: Ben Wolsieffer <benwolsieffer@gmail.com>
---
 sound/soc/codecs/sgtl5000.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/sound/soc/codecs/sgtl5000.c b/sound/soc/codecs/sgtl5000.c
index 60764f6201b1..d9f9f441d4e7 100644
--- a/sound/soc/codecs/sgtl5000.c
+++ b/sound/soc/codecs/sgtl5000.c
@@ -260,6 +260,18 @@ static SOC_ENUM_SINGLE_DECL(dapmix_enum,
 static const struct snd_kcontrol_new dapmix_mux =
 SOC_DAPM_ENUM("DAP MIX Mux", dapmix_enum);
 
+/* input sources for I2S output */
+static const char *i2s_out_mux_text[] = {
+	"ADC", "I2S", "Rsvrd", "DAP"
+};
+
+static SOC_ENUM_SINGLE_DECL(i2s_out_enum,
+			    SGTL5000_CHIP_SSS_CTRL, SGTL5000_I2S_OUT_SEL_SHIFT,
+			    i2s_out_mux_text);
+
+static const struct snd_kcontrol_new i2s_out_mux =
+SOC_DAPM_ENUM("I2S Output Mux", i2s_out_enum);
+
 
 static const struct snd_soc_dapm_widget sgtl5000_dapm_widgets[] = {
 	SND_SOC_DAPM_INPUT("LINE_IN"),
@@ -281,7 +293,7 @@ static const struct snd_soc_dapm_widget sgtl5000_dapm_widgets[] = {
 	SND_SOC_DAPM_MUX("DAP Mux", SGTL5000_DAP_CTRL, 0, 0, &dap_mux),
 	SND_SOC_DAPM_MUX("DAP MIX Mux", SGTL5000_DAP_CTRL, 4, 0, &dapmix_mux),
 	SND_SOC_DAPM_MIXER("DAP", SGTL5000_CHIP_DIG_POWER, 4, 0, NULL, 0),
-
+	SND_SOC_DAPM_MUX("I2S Output Mux", SND_SOC_NOPM, 0, 0, &i2s_out_mux),
 
 	/* aif for i2s input */
 	SND_SOC_DAPM_AIF_IN("AIFIN", "Playback",
@@ -306,7 +318,6 @@ static const struct snd_soc_dapm_route sgtl5000_dapm_routes[] = {
 	{"Capture Mux", "MIC_IN", "MIC_IN"},	/* mic_in --> adc_mux */
 
 	{"ADC", NULL, "Capture Mux"},		/* adc_mux --> adc */
-	{"AIFOUT", NULL, "ADC"},		/* adc --> i2s_out */
 
 	{"DAP Mux", "ADC", "ADC"},		/* adc --> DAP mux */
 	{"DAP Mux", NULL, "AIFIN"},		/* i2s --> DAP mux */
@@ -321,6 +332,11 @@ static const struct snd_soc_dapm_route sgtl5000_dapm_routes[] = {
 	{"Digital Input Mux", NULL, "DAP"},	/* dap --> audio mux */
 	{"DAC", NULL, "Digital Input Mux"},	/* audio mux --> dac */
 
+	{"I2S Output Mux", "ADC", "ADC"},	/* adc --> i2s_out mux */
+	{"I2S Output Mux", NULL, "AIFIN"},	/* i2s --> i2s_out mux */
+	{"I2S Output Mux", NULL, "DAP"},	/* dap --> i2s_out mux */
+	{"AIFOUT", NULL, "I2S Output Mux"},	/* i2s_out mux --> i2s_out */
+
 	{"Headphone Mux", "DAC", "DAC"},	/* dac --> hp_mux */
 	{"LO", NULL, "DAC"},			/* dac --> line_out */
 
-- 
2.19.1

